polkadot/node/network/statement-distribution/src/v2/mod.rs

```diff
@@ -2139,47 +2139,52 @@ async fn provide_candidate_to_grid<Context>(
 
 // Utility function to populate per relay parent `ParaId` to `GroupIndex` mappings.
 fn determine_groups_per_para(
-	availability_cores: &[CoreState],
-	group_rotation_info: &GroupRotationInfo,
-	maybe_claim_queue: &Option<ClaimQueueSnapshot>,
+	availability_cores: Vec<CoreState>,
+	group_rotation_info: GroupRotationInfo,
+	maybe_claim_queue: Option<ClaimQueueSnapshot>,
 	max_candidate_depth: usize,
 ) -> HashMap<ParaId, Vec<GroupIndex>> {
+	let n_cores = availability_cores.len();
+
 	// Determine the core indices occupied by each para at the current relay parent. To support
 	// on-demand parachains we also consider the core indices at next block if core has a candidate
 	// pending availability.
-	let para_core_indices = availability_cores.iter().enumerate().filter_map(|(index, core)| {
-		match core {
-			CoreState::Scheduled(scheduled_core) =>
-				Some((scheduled_core.para_id, CoreIndex(index as u32))),
-			CoreState::Occupied(occupied_core) => {
-				if max_candidate_depth >= 1 {
-					// Use claim queue if available, or fallback to `next_up_on_available`
-					let maybe_scheduled_core = match maybe_claim_queue {
-						Some(claim_queue) => {
-							// What's up next on this core ?
-							fetch_next_scheduled_on_core(claim_queue, CoreIndex(index as u32))
-						},
-						None => {
-							// Runtime doesn't support claim queue runtime api. Fallback to
-							// `next_up_on_available`
-							occupied_core.next_up_on_available.clone()
-						},
-					};
-
-					maybe_scheduled_core
-						.map(|scheduled_core| (scheduled_core.para_id, CoreIndex(index as u32)))
-				} else {
-					None
-				}
-			},
-			CoreState::Free => None,
-		}
-	});
+	let para_core_indices: Vec<_> = if let Some(claim_queue) = maybe_claim_queue {
+		claim_queue
+			.keys()
+			.filter_map(|core_index| {
+				let Some(scheduled_core) = fetch_next_scheduled_on_core(&claim_queue, *core_index)
```

Comment:
Ok, first (not introduced here): I find the name of this function vastly misleading as it does not fetch anything.

Second, why are we iterating only keys of the claim queue to fetch the actual element for each key, instead of iterating key/values to begin with?

Reply:
I wanted to reuse the fn, but you are right, it is actually useless in this case. A wrapper for this BTreeMap with a method like `iter_claims(depth: BlockNumber)` returning an iterator over `(core_index, para)` so we can lookup as far into the future as the lookahead param/ async backing settings allow.

Reply:
@tdimitrov wdyt ?

Reply:
https://github.com/paritytech/polkadot-sdk/pull/3950